cpp/src/io/utilities/datasource.cpp

```diff
@@ -63,20 +76,19 @@ class file_source : public datasource {
                      rmm::cuda_stream_view stream) override
   {
     CUDF_EXPECTS(supports_device_read(), "Device reads are not supported for this file.");
-
-    auto const read_size = std::min(size, _file.size() - offset);
-    return _cufile_in->read(offset, read_size, dst, stream);
+    return device_read_async(offset, size, dst, stream).get();
```

Comment:
The exception is also duplicated and can be removed

Reply:
Fixed